### ISSUE TITLE
[Feature] Add possibility to provide custom properties for custom sidebar

### ISSUE THREAD

**Issue by Nxtivision** [ng-doc/ng-doc#185]
### Description

Hello!

I've followed the guide to create my custom Sidebar and after extracting components from your Github, I've managed to get my own sidebar. 
However, I thought I had access to all properties of NgDocPage but it is more restricted than planned.

Could you provide a property where we can put evey custom data for our custom sidebar ?

### Proposed solution

```typescript
export interface NgDocNavigation extends NgDocBaseEntity {
  expandable?: boolean;
  /** Determines whether the category should be expanded by default */
  expanded?: boolean;
  /** Children of the navigation item */
  children?: NgDocNavigation[];
  /** The route of the entity (current sourceFileFolder name by default) */
  route: string;

  /** Other things you want to integrate inside the custom components **/
  custom: {
    sidebar: Record<string, any>;
    navbar: Record<string, any>;
    // and so on...
  }
}
```

Here is an example of what I would like in order to customize properly the sidebar :)

### Alternatives considered

Using `data` from NgDocPage, but it is not for this use case

**Comment by skoropadas** [comment]
But what are you trying to achieve?

**Comment by Nxtivision** [comment]
Here is an example of what I try to achieve:

My client would like to see a tag, an indicator directly visible in the sidebar when I meet a certain condition.
So I've created the custom sidebar and retrieved the necessary information from `NG_DOC_CONTEXT`

```typescript
export class CustomSidebarComponent {
  constructor(@Inject(NG_DOC_CONTEXT) private context: NgDocContext) {
    // `context.navigation` contains the list of categories and pages of documentation
    console.log(context);
  }

  getNavigation(nav?: NgDocNavigation): NgDocNavigation[] {
    return nav ? nav.children ?? [] : this.context.navigation;
  }
}
```

However, the only information available are:

![2024-07-15_13h00_46](https://github.com/user-attachments/assets/bd0310a4-6b78-45c0-b50b-fd7c58a4882f)

What I would like is a property where I can put custom information that I can use inside the sidebar :)

Here is an example of my custom sidebar item:

```html
<a
  class="ng-doc-sidebar-link"
  [routerLink]="[item?.route ?? '']"
  routerLinkActive="active">
  <ng-doc-dot></ng-doc-dot>
  <span ng-doc-text [absoluteContent]="true">
    {{ item?.title }}
  </span>
  @if (item?.custom['sidebar']['tag']) {
    <my-tag></my-tag>
  }
</a>
```

That means, inside the `NgDocPage`, I could be able to customize the sidebar or other parts of the UI:

```typescript
const ButtonIconPage: NgDocPage = {
  title: `Button Icon`,
  mdFile: './index.md',
  category: OneThemeCategory,
  demos: {
    DemoButtonIconComponent,
    DemoButtonIconDisabledComponent,
  },
  custom: {
    sidebar: {
      tag: true,
    }
  },
};

export default ButtonIconPage;
```

What do you think about this? 😄

**Comment by skoropadas** [comment]
I'm still thinking about solution. Adding an additional field in `NgDocPage` would require importing each page at the `NG_DOC_CONTEXT` level, leading to a loss of tree-shaking and inclusion in the main application bundle.

In theory, I could compile your file separately, removing all fields except `custom`, so that Angular components and other entities do not end up in `NG_DOC_CONTEXT`. However, this seems too complicated and time-consuming.

It all depends on the specific data you want to transmit. If you don't need JavaScript, theoretically we could use JSDoc tags, for example:

```ts
/**
  @mytag
  @myCustomTitle MyCustomTitle
*/
const ButtonIconPage: NgDocPage = {
  title: `Button Icon`,
  mdFile: './index.md',
  category: OneThemeCategory,
  demos: {
    DemoButtonIconComponent,
    DemoButtonIconDisabledComponent,
  },
};

export default ButtonIconPage;
```

Then in the `NgDocNavigation` object, you would obtain the value like this:

```js
{title: 'Button Icon' metadata: {mytag: true, myCustomTitle: 'MyCustomTitle'}}
```

What do you think?

**Comment by skoropadas** [comment]
Ok, I've added this functionality to the NgDoc 18, I think it works well especially with new API,

So e.g. now in NgDoc 18 you can display badges in sidebar using JSDoc tags

```ts
/**
 * This section describes how you can configure certain functions of the library.
 *
 * @status:success NEW OPTIONS
 */
const Configuration: NgDocPage = {
  title: `Configuration`,
  mdFile: './index.md',
  category: GettingStartedCategory,
  order: 3,
};

export default Configuration;
```

<img width="647" alt="image" src="https://github.com/user-attachments/assets/d59f0927-5aca-421a-be48-f7aacf9dce82">

And if you are creating your own sidebar you will have an access to this metadata in `NgDocNavigation` object:

```ts
{
    "title": "Configuration",
    "route": "/docs/getting-started/configuration",
    "order": 3,
    "hidden": false,
    "metadata": {
        "description": "<p>This section describes how you can configure certain functions of the library.</p>",
        "tags": {
            "status": [
                ":success NEW OPTIONS"
            ]
        }
    }
}
```

**Comment by akhileshThapliyal** [comment]
> Ok, I've added this functionality to the NgDoc 18, I think it works well especially with new API,
> 
> So e.g. now in NgDoc 18 you can display badges in sidebar using JSDoc tags
> 
> ```ts
> /**
>  * This section describes how you can configure certain functions of the library.
>  *
>  * @status:success NEW OPTIONS
>  */
> const Configuration: NgDocPage = {
>   title: `Configuration`,
>   mdFile: './index.md',
>   category: GettingStartedCategory,
>   order: 3,
> };
> 
> export default Configuration;
> ```
> 
> <img alt="image" width="647" src="https://private-user-images.githubusercontent.com/20700969/351045244-d59f0927-5aca-421a-be48-f7aacf9dce82.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIxNzE3MDIsIm5iZiI6MTcyMjE3MTQwMiwicGF0aCI6Ii8yMDcwMDk2OS8zNTEwNDUyNDQtZDU5ZjA5MjctNWFjYS00MjFhLWJlNDgtZjdhYWNmOWRjZTgyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI4VDEyNTY0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjYzdjMGIzNzVlNzAyZGM0OTkyNDZiYzY3ZjYxM2VjMmJiMGQ4NGI4ZDk3MTQ5YjYyNzMwZjM5Y2ZmZTM0ZDAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.c3AFh2L97xDL2Ex_B5cxHhxJNYr_ZbEnfIBOid5cG0I">
> And if you are creating your own sidebar you will have an access to this metadata in `NgDocNavigation` object:
> 
> ```ts
> {
>     "title": "Configuration",
>     "route": "/docs/getting-started/configuration",
>     "order": 3,
>     "hidden": false,
>     "metadata": {
>         "description": "<p>This section describes how you can configure certain functions of the library.</p>",
>         "tags": {
>             "status": [
>                 ":success NEW OPTIONS"
>             ]
>         }
>     }
> }
> ```

Thank you @Nxtivision and @skoropadas, I was also looking for a similar feature. Let me check if this could also help us achieve a custom sidebar navigation where we want to pass some metadata to render the sidebar which fits our design requirements. I will keep you informed.

<img width="196" alt="image" src="https://github.com/user-attachments/assets/d6859f5b-4fef-45e4-8b82-b636fab5a4a8">

**Comment by skoropadas** [comment]
@akhileshThapliyal yes it will help you to do that. You can use any tag you want. In the example above I used `status` tag which doesn't exist in the jsdoc specification, so you can use e.g. `image` tag and put the image name after it and handle its value in the sidebar.

**Comment by akhileshThapliyal** [comment]
> @akhileshThapliyal yes it will help you to do that. You can use any tag you want. In the example above I used `status` tag which doesn't exist in the jsdoc specification, so you can use e.g. `image` tag and put the image name after it and handle its value in the sidebar.

@skoropadas, so metadata attribute type is of Any right?

**Comment by skoropadas** [comment]
@akhileshThapliyal it's basically string, you can see it [here](https://github.com/ng-doc/ng-doc/blob/feat%2Fng-doc-18/libs%2Fcore%2Finterfaces%2Fjs-doc-metadata.ts)

**Comment by skoropadas** [comment]
Released in NgDoc 18
